server/service/service_osquery.go

```
@@ -420,7 +421,23 @@ func (svc service) ingestDistributedQuery(host kolide.Host, name string, rows []
 
 	err = svc.resultStore.WriteResult(res)
 	if err != nil {
-		return osqueryError{message: "writing results: " + err.Error()}
+		nErr, ok := err.(pubsub.NoSubscriber)
```

This is not a big point of contention, but I feel that if you're going to import, a type check against the concrete type works just as well. 

```
if _,ok := err.(interface{NoSubscriber() bool};ok { }``` would achieve the same result and avoid coupling the code here to a specific package/import.

---

I think it makes sense to make the type assertion directly against the interface. In either case, there is a coupling between this package and the pubsub package via this interface. Using the explicit reference to `pubsub.NoSubscriber`, if that interface were to be removed (or changed in certain ways) in `pubsub`, we would generate a compile time error at this point. With your proposal, we could silently fail due to any changes in that package. I would prefer to take some compiler help over none.

---

ok I mostly agree. In this case I'd argue that the error should be something like 

```
type Error interface {
    error
    NoSubscriber() bool
    Future() bool
}
``` 
similar to how `net.Error()` in stdlib is described. 

https://blog.golang.org/error-handling-and-go

---

I think we're in agreement.

Just to clarify, does `Future()` in this example represent some future other method we might add to the interface? Based on your comment above, I am prepared to change the current interface to:

```
type NoSubscriber interface {
    error
    NoSubscriber() bool
}
```

---

Yeah, Future() is a made up error. 

 I was suggesting naming the interface `pubsub.Error` instead of `NoSubscriber`.
It would make sense to do that to allow for adding other methods in the future.

---

Ah, interesting. I'm fine with doing that.